File: php/examples/workflow_wasm.php
<?php

require_once __DIR__ . '/../vendor/autoload.php';

use NativeBPM\Client\Builder\Workflow;

echo "=== NativeBPM PHP SDK: Workflow as Code ===\n";

// 1. Build workflow dynamically using Workflow as Code Fluent API
echo "🔨 Building workflow dynamically using Fluent API...\n";
$workflow = new Workflow("native-demo", "Workflow as Code");

// Chain starting from the start event
$workflow->startEvent("start")
    ->next("gateway");

// Configure exclusive gateway with transition conditions and default flow
$workflow->exclusiveGateway("gateway", "Urgency Gateway")
    ->condition("reviewOrder", '${isUrgent == true}')
    ->defaultValue("notifyCustomer");

// Add standard Service Task. Since no `->wasm(...)` call is made,
// this task is executed natively as a topic-based worker queue execution.
$workflow->serviceTask("notifyCustomer", "Send Confirmation Email", "email_topic")
    ->next("end");

// Add User Task and chain to the end event
$workflow->userTask("reviewOrder", "Review Order Details")
    ->assignee("sales_representative")
    ->next("end");

$workflow->endEvent("end", "Process Finished");

$bpmnXml = $workflow->buildXML();
echo "✓ Successfully compiled native workflow AST to BPMN 2.0 XML.\n";

// Print a snippet of generated XML
if (strlen($bpmnXml) > 300) {
    echo "XML snippet:\n" . substr($bpmnXml, 0, 300) . "...\n";
} else {
    echo "XML output:\n" . $bpmnXml . "\n";
}

File: rust/examples/workflow_wasm.rs
use nativebpm_client::builder::Workflow;
use nativebpm_client::apis::{configuration, default_api};

#[tokio::main]
async fn main() {
    println!("=== NativeBPM Rust SDK: Workflow as Code ===");

    // 1. Build workflow as code using Fluent API method chaining
    let mut workflow = Workflow::new("native-demo", "Workflow as Code");

    // Chain starting from the start event
    workflow.start_event("start")
        .connect_to("gateway")
        .exclusive_gateway("gateway", "Urgency Gateway")
        .condition("reviewOrder", "${isUrgent == true}")
        .default("notifyCustomer")
        .builder()
        .service_task("notifyCustomer", "Send Confirmation Email", "email_topic")
        .connect_to("end")
        .user_task("reviewOrder", "Review Order Details")
        .assignee("sales_representative")
        .connect_to("end")
        .end_event("end", "Process Finished");

    // Compile the workflow AST to standard BPMN 2.0 XML using the embedded Go engine
    let bpmn_xml = match workflow.build_xml() {
        Ok(xml) => {
            println!("✓ Successfully compiled native workflow AST to BPMN 2.0 XML.");
            xml
        }
        Err(e) => {
            eprintln!("✗ Failed to compile workflow: {}", e);
            return;
        }
    };

    // 2. Deploy and start process definition using the REST API client
    let mut config = configuration::Configuration::new();
    config.base_path = "http://localhost:8080".to_string();
    config.bearer_access_token = Some("test-bearer-token".to_string());

    println!("\nDeploying to NativeBPM engine...");
    
    // In actual usage, you would write the XML to a file to deploy via multipart/form-data
    let bpmn_path = std::env::temp_dir().join("native-demo.bpmn");
    if std::fs::write(&bpmn_path, bpmn_xml).is_ok() {
        match default_api::deploy_definition(&config, Some(bpmn_path)).await {
            Ok(definition) => {
                println!("✓ Deployed process definition (hash: {:?})", definition.hash);

                // Start a process instance
                let mut variables = std::collections::HashMap::new();
                variables.insert("customerEmail".to_string(), serde_json::json!("customer@example.com"));
                variables.insert("isUrgent".to_string(), serde_json::json!(true));

                let start_request = nativebpm_client::models::StartInstanceRequest {
                    instance_id: None,
                    business_key: Some("order-5541".to_string()),
                    variables: Some(variables),
                };

                match default_api::start_instance(&config, "native-demo", Some(start_request)).await {
                    Ok(instance) => {
                        println!("✓ Started process instance ID: {} (completed: {})", instance.id, instance.completed);
                    }
                    Err(e) => {
                        println!("Note: Failed to start process instance: {:?}", e);
                    }
                }
            }
            Err(e) => {
                println!("Note: Local API Engine deployment skipped (ensure local server is running on :8080). Details: {:?}", e);
            }
        }
    }
}
